File: python-rag-agent/app/rag/splitter.py
"""文档切分:把 topic 的 learningCards 拆成 chunk。这是 RAG 链路的第二环。

核心设计——按卡片类型分流:
- explain/interviewAnswer 长文用 RecursiveCharacterTextSplitter 切;
- checklist/compareTable/code 短卡整张入库(切了反而破坏结构);
- diagram 的 content 是 mermaid 源码,整张入库(fallback 文本也存)。

为什么这样分:本知识库以"短而结构化"的卡片为主,整卡语义完整,强行切分会
破坏列表/表格/代码的完整性;只有长段落才需要切。因此切分前先按类型判断。

chunk_size 500 / overlap 80 是经验值:多数短问答能整条入库,长答案才切,
overlap 80 保证边界语义不断裂。

为避免 langchain 依赖过重,这里实现一个等价的最简递归切分器;
若 langchain 可用则优先用官方实现(深挖可对比)。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from app.config import get_settings
from app.rag.loader import Topic

# 整张入库的 card 类型(短/结构化,切了破坏语义):清单、对比表、代码、图示
WHOLE_CARD_TYPES = {"checklist", "compareTable", "code", "diagram"}
# 需要切分的 card 类型(长文):讲解、面试参考答案
SPLIT_CARD_TYPES = {"explain", "interviewAnswer"}


@dataclass
class Chunk:
    """切分后的单个文本片段。

    text     片段正文,用于向量化与 BM25 索引;
    metadata 溯源与过滤信息(topic_id/domain/title/tags/difficulty/card_type),
             检索命中后据此回填"来源",实现答案可溯源。
    """

    text: str
    metadata: dict[str, Any] = field(default_factory=dict)


def _split_recursive(text: str, chunk_size: int, chunk_overlap: int) -> list[str]:
    """最简递归切分:按段落 -> 句子 -> 字符 三级降级。

    与 langchain RecursiveCharacterTextSplitter 思路一致,但无外部依赖。

    三步:
      1. 依次尝试更细的分隔符(段落->换行->句号->空格),把文本拆到每段都不超
         chunk_size 为止。优先在语义边界(段落/句子)断开,尽量不把句子切碎。
      2. 仍有超长片段(如没有任何标点的长串)时按字符硬切兜底。
      3. 把碎片重新拼装回接近 chunk_size 的块,并在相邻块间保留 chunk_overlap
         字符的重叠,避免关键信息恰好落在切分边界而丢失。
    """
    if len(text) <= chunk_size:
        return [text] if text.strip() else []

    # 分隔符优先级:从最"语义完整"的段落级,逐级降到字符级。
    seps = ["\n\n", "\n", "。", ".", "!", "?", ";", ";", " "]
    pieces = [text]
    for sep in seps:
        new_pieces = []
        for p in pieces:
            if len(p) <= chunk_size:
                new_pieces.append(p)
            else:
                new_pieces.extend(s for s in p.split(sep) if s)
        pieces = new_pieces
        if all(len(p) <= chunk_size for p in pieces):
            break

    # 仍有超长 piece:按字符硬切(最后的兜底)
    final: list[str] = []
    for p in pieces:
        if len(p) <= chunk_size:
            final.append(p)
        else:
            for i in range(0, len(p), chunk_size):
                final.append(p[i : i + chunk_size])
    pieces = final

    # 合并成 chunk_size 大小,带 overlap
    out: list[str] = []
    buf = ""
    for p in pieces:
        if not p:
            continue
        if buf and len(buf) + len(p) + 1 > chunk_size:
            out.append(buf.strip())
            # 新块以旧块尾部 chunk_overlap 个字符开头,实现相邻块重叠,避免边界语义断裂。
            buf = buf[-chunk_overlap:] + p
        else:
            buf = (buf + p) if buf else p
    if buf.strip():
        out.append(buf.strip())
    return out


def split_topic(topic: Topic, *, chunk_size: int | None = None, chunk_overlap: int | None = None) -> list[Chunk]:
    """把单个 topic 的 learningCards 切成 chunk,带 metadata(topic_id/domain/title/tags/difficulty/card_type)。

    参数:
      chunk_size/chunk_overlap 缺省时取配置默认值(rag_chunk_size/rag_chunk_overlap)。
    逻辑:
      遍历每张学习卡片,按类型分流:短结构化卡整张入库、长文递归切分;
      最后把 summary 也单独作为一个 chunk 入库。
    返回:Chunk 列表(空卡跳过)。
    """
    s = get_settings()
    cs = chunk_size or s.rag_chunk_size
    co = chunk_overlap or s.rag_chunk_overlap

    chunks: list[Chunk] = []
    # 所有 chunk 共享的溯源 metadata:检索命中后据此回填来源。
    base_meta = {
        "topic_id": topic.id,
        "domain": topic.domain,
        "category": topic.category,
        "title": topic.title,
        "tags": topic.tags,
        "difficulty": topic.difficulty,
    }

    for card in topic.learning_cards or []:
        ctype = card.get("type", "explain")
        title = card.get("title", "")
        content = card.get("content", "")
        if not content:
            continue  # 空卡片跳过

        meta = {**base_meta, "card_type": ctype, "card_title": title}

        if ctype in WHOLE_CARD_TYPES:
            # 短/结构化卡整张入库,不切分(切了破坏清单/表格/代码结构)。
            chunks.append(Chunk(text=content, metadata=meta))
            # diagram 的 fallback 文本也入库(便于检索)
            fb = card.get("fallback")
            if ctype == "diagram" and fb:
                chunks.append(Chunk(text=fb, metadata={**meta, "card_title": f"{title}(文本版)"}))
        elif ctype in SPLIT_CARD_TYPES:
            # 长文(explain/interviewAnswer)递归切分。
            for piece in _split_recursive(content, cs, co):
                chunks.append(Chunk(text=piece, metadata=meta))
        else:
            # 未知类型保守整张入库。
            chunks.append(Chunk(text=content, metadata=meta))

    # summary 也作为一个 chunk(短摘要,检索标题/概述时命中)
    if topic.summary:
        chunks.append(Chunk(text=topic.summary, metadata={**base_meta, "card_type": "summary", "card_title": "摘要"}))

    return chunks


def split_topics(topics: list[Topic], **kwargs) -> list[Chunk]:
    """批量切分多个 topic,聚合为单个 chunk 列表(供 ingest 一次性入库)。"""
    out: list[Chunk] = []
    for t in topics:
        out.extend(split_topic(t, **kwargs))
    return out

File: java-ai-service/src/main/java/com/nontracey/aiservice/rag/Splitter.java
package com.nontracey.aiservice.rag;

import com.nontracey.aiservice.dto.Dtos.Topic;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * 文档切分器(RAG 的切块环节):把 topic 的 learningCards 拆成可检索的 {@link Chunk}。
 *
 * <p><b>架构位置</b>:rag 模块,位于 Loader 之后、VectorStoreService 之前。产出的 chunk
 * 携带丰富 metadata(topic_id/title/card_type 等),既供检索来源标注,也供租户过滤。
 *
 * <p><b>切块策略(与 B 同策略)</b>:
 * <ul>
 *   <li>短卡片(checklist / compareTable / code / diagram)整张入库,避免破坏结构完整性。</li>
 *   <li>长文(explain / interviewAnswer)按 {@link #CHUNK_SIZE} 递归切分,带 {@link #CHUNK_OVERLAP} 重叠。</li>
 *   <li>topic.summary 作为独立 summary chunk 入库。</li>
 * </ul>
 *
 * <p><b>同构映射</b>:对应 B 项目 {@code app/rag/splitter.py}。
 */
@Service
public class Splitter {

    /** 整张入库、不切分的卡片类型(结构化内容,切开会破坏语义)。 */
    private static final Set<String> WHOLE_CARD_TYPES = Set.of("checklist", "compareTable", "code", "diagram");
    /** 需要按长度递归切分的卡片类型(长文本)。 */
    private static final Set<String> SPLIT_CARD_TYPES = Set.of("explain", "interviewAnswer");

    /** 单 chunk 目标最大字符数。 */
    public static final int CHUNK_SIZE = 500;
    /** 相邻 chunk 重叠字符数,避免语义在边界被截断。 */
    public static final int CHUNK_OVERLAP = 80;

    /**
     * 把单个 topic 的全部 learningCards(及 summary)切成 chunk 列表。
     *
     * @param topic 知识库条目
     * @return chunk 列表(每个都带 topic 级 + 卡片级 metadata)
     */
    public List<Chunk> splitTopic(Topic topic) {
        List<Chunk> out = new ArrayList<>();
        // topic 级公共 metadata,所有 chunk 共享,供检索来源标注与过滤
        Map<String, Object> base = new HashMap<>();
        base.put("topic_id", topic.id());
        base.put("domain", topic.domain());
        base.put("category", topic.category());
        base.put("title", topic.title());
        base.put("tags", topic.tags());
        base.put("difficulty", topic.difficulty());

        for (Map<String, Object> card : topic.learningCards()) {
            String ctype = (String) card.getOrDefault("type", "explain");
            String title = (String) card.getOrDefault("title", "");
            String content = (String) card.getOrDefault("content", "");
            if (content == null || content.isBlank()) continue;

            // 卡片级 metadata:在 topic 级基础上追加 card_type / card_title
            Map<String, Object> meta = new HashMap<>(base);
            meta.put("card_type", ctype);
            meta.put("card_title", title);

            if (WHOLE_CARD_TYPES.contains(ctype)) {
                // 结构化短卡整张入库,避免切碎破坏语义
                out.add(new Chunk(content, meta));
                // diagram 卡片若带纯文本 fallback,额外入一份文本版,便于纯文本检索
                if ("diagram".equals(ctype) && card.get("fallback") instanceof String fb && !fb.isBlank()) {
                    Map<String, Object> m2 = new HashMap<>(meta);
                    m2.put("card_title", title + "(文本版)");
                    out.add(new Chunk(fb, m2));
                }
            } else if (SPLIT_CARD_TYPES.contains(ctype)) {
                // 长文本递归切分,带 overlap
                for (String piece : splitRecursive(content, CHUNK_SIZE, CHUNK_OVERLAP)) {
                    out.add(new Chunk(piece, meta));
                }
            } else {
                // 未知类型兜底:整张入库
                out.add(new Chunk(content, meta));
            }
        }
        // topic 摘要单独成一个 chunk,便于概览类问题命中
        if (topic.summary() != null && !topic.summary().isBlank()) {
            Map<String, Object> m = new HashMap<>(base);
            m.put("card_type", "summary");
            m.put("card_title", "摘要");
            out.add(new Chunk(topic.summary(), m));
        }
        return out;
    }

    /**
     * 批量切分多个 topic(ingest 流程调用)。
     *
     * @param topics topic 列表
     * @return 全部 chunk(顺序与 topic 顺序一致)
     */
    public List<Chunk> splitAll(List<Topic> topics) {
        List<Chunk> out = new ArrayList<>();
        for (Topic t : topics) out.addAll(splitTopic(t));
        return out;
    }

    /**
     * 最简递归切分(与 B 同策略)。
     *
     * <p>三步走:
     * <ol>
     *   <li>按分隔符优先级({@code \n\n > \n > 句号/点 > 空格})逐级拆分,尽量在自然语义边界断开;
     *       一旦所有片段都不超过 size 就提前结束。</li>
     *   <li>对仍超长的片段硬切(每 size 一段)兜底,保证没有超大 chunk。</li>
     *   <li>把相邻小片段合并到接近 size 的块,并在块之间保留 overlap 个字符的重叠,
     *       避免关键语义正好被截断在边界。</li>
     * </ol>
     *
     * @param text    原始长文本
     * @param size    单块目标最大长度
     * @param overlap 相邻块重叠字符数
     * @return 切分后的文本片段列表
     */
    private List<String> splitRecursive(String text, int size, int overlap) {
        if (text.length() <= size) return List.of(text);
        String[] seps = {"\n\n", "\n", "。", ".", "!", "?", ";", " "};
        List<String> pieces = new ArrayList<>(List.of(text));
        for (String sep : seps) {
            List<String> next = new ArrayList<>();
            for (String p : pieces) {
                if (p.length() <= size) next.add(p);
                else {
                    for (String s : p.split(java.util.regex.Pattern.quote(sep))) {
                        if (!s.isEmpty()) next.add(s);
                    }
                }
            }
            pieces = next;
            if (pieces.stream().allMatch(p -> p.length() <= size)) break;
        }
        // 兜底硬切
        List<String> fin = new ArrayList<>();
        for (String p : pieces) {
            if (p.length() <= size) fin.add(p);
            else for (int i = 0; i < p.length(); i += size) fin.add(p.substring(i, Math.min(p.length(), i + size)));
        }
        // 合并带 overlap
        List<String> out = new ArrayList<>();
        StringBuilder buf = new StringBuilder();
        for (String p : fin) {
            if (buf.length() > 0 && buf.length() + p.length() + 1 > size) {
                out.add(buf.toString().trim());
                buf = new StringBuilder(buf.substring(Math.max(0, buf.length() - overlap))).append(p);
            } else {
                buf.append(p);
            }
        }
        if (!buf.toString().trim().isEmpty()) out.add(buf.toString().trim());
        return out;
    }

    /**
     * 切块后的最小检索/入库单元。
     *
     * @param text     chunk 文本(用于 embedding 与 BM25 索引)
     * @param metadata 元数据(topic_id/title/card_type/tags 等,供来源标注与租户过滤)
     */
    public record Chunk(String text, Map<String, Object> metadata) {}
}
